### PR TITLE
Improve rice dependency resolution and installation support

### PR DESCRIPTION
**Pull request by 2maz**
These commits provide the basis for a proper project-specific installation of ruby extensions to avoid conflicts and allow for the usage of a debian-package of rice.

### CODE REVIEW

**Comment by 2maz** [comment]
any comments from @rock-core/core ?

File: modules/RockRuby.cmake [hunk 1]
```diff
@@ -21,6 +21,10 @@
 # endif()
 #
 # 
+# If installed to a target directory set the following to true to
+# make it project specific
+# set(RUBY_USE_PROJECT_INSTALL_PREFIX TRUE)
```
Comment (doudou) [inline modules/RockRuby.cmake]: I'm really against having yet-another-flag. Properly layout the bindings folder to match a gem-like folder structure (with bin/ lib/ and test/) and do `rock_add_ruby_package(lib/my_package_name)`

Example: https://github.com/rock-core/drivers-transformer/blob/master/ruby/CMakeLists.txt

**Comment by doudou** [comment]
Who is installing the pkg-config file for Rice again ? I'm missing something here ...

**Comment by 2maz** [comment]
Sorry its not mentioned in the commit message. That is particularly targeting the rock-debian packing which packages the rice gem and dependant packages will rely on pkg-config to find the library.

**Comment by doudou** [comment]
> dependant packages will rely on pkg-config to find the library.

Debian and Ubuntu ship with rubygem-integration, which allows finding the OS gems through the 'normal' rubygem process. Isn't that good enough ?

I think we should avoid having a specific mechanism to resolve gems when packaging.

**Comment by 2maz** [comment]
For rice we have to find the library in particular and currently we use FindGem.cmake for finding the Gem (which is not an OS package) and the library. What are the details of the 'normal' rubygem process with rubygem-integration if I want to find the static library of rice? It seems the currently used mechanism ('gem contents') does not work for OS-packages, but you can correct me here.

The easiest and most robust way for me was to install a pc file along with the (now Rock-specific) package, which aligns fine with the Rock infrastructure which relies on pkg-config for dependency resolution to main parts.

File: modules/RockRuby.cmake (hunk 1 above)
Comment (2maz) [inline modules/RockRuby.cmake]: gui/vizkit is the particular example triggering that addition -- here rock_ruby_rice_extension macro is being used, but that will require adapting the RUBY_EXTENSION_INSTALL_DIR or RUBY_EXTENSION_LIBRARY_DIR directly to achieve the same result

**Comment by doudou** [comment]
`gem contents` indeed does not seem to work. However, `gem which` seems to. And AFAIK, `gem which` works for RUBYLIB in general (does not require the file to be within a rubygem)

```
[~]% sudo apt install ruby-nokogiri
[~]% /usr/bin/gem which nokogiri/nokogiri.so
/usr/lib/x86_64-linux-gnu/ruby/vendor_ruby/2.1.0/nokogiri/nokogiri.so
[~/dev/flat_fish]% gem which roby.rb
/home/doudou/dev/flat_fish/tools/roby/lib/roby.rb
```

**Comment by 2maz** [comment]
I will have a look into using `gem which` as fallback - however, do you think that using 'gem which' will be more robust? 
There will be more guessing involved compared to pkgconfig and FindGem already provides the option to search for  files? 
Additionally, rice gem contents look already more like a general library with headers and static lib and on top of that I had and have to patch it anyhow to make it install properly.

File: modules/RockRuby.cmake (hunk 1 above)
Comment (doudou) [inline modules/RockRuby.cmake]: I misunderstood the patch - did not get that it was involving the extensions, not the ruby parts.

I would go for having the same default behaviour in CMake that in the gems ... that is installing the extension under <project_name>/ always.

As far as I can see, that change would not impact anyone, as nobody uses the install except the binary packaging anyways, or ?

**Comment by doudou** [comment]
I think that anything that requires less source patching will be more robust.

Could you point me to the patch you need on it to make it work ?

**Comment by 2maz** [comment]
Have a look at https://github.com/2maz/deb_patches/tree/master/rice.
Note however that the pkconfig file will receive an additional first line defining the rock_install_dir.
That is part of the automated package generation. 

The general process that is applied for gems like rice:

gem2tgz rice
# -p does rename the gem and thus defines the name for the resulting debian package
dh-make-ruby rice-2.1.0.tar.gz -p rock-master-ruby-rice
cd rock-master-ruby-rice-2.1.0
... apply the changes: the given 'overlay' in 2maz/deb_patches will be replacing any existing files
dpkg-buildpackage -us -uc
...
dpkg --contents ../rock-master-ruby-rice_2.1.0-1_amd64.deb

File: modules/RockRuby.cmake (hunk 1 above)
Comment (2maz) [inline modules/RockRuby.cmake]: Ok, I reecked and it might indeed be the wrong fix - so for vizkit the current situation is the following: the extension is required as
    require 'vizkit/vizkittypelib'

So I guess that should be fixed to "require 'vizkittypelib' then" and same for TypelibQtAdapter

Comment (2maz) [inline modules/RockRuby.cmake]: Updated gui-vizkit#54 accordingly

Comment (doudou) [inline modules/RockRuby.cmake]: Misunderstanding. I think that require 'vizkit/vizkittypelib' is the right require, so I was suggesting to change the CMake code to always (non-optionally) install into LIBDIR/<project_name>/

Comment (2maz) [inline modules/RockRuby.cmake]: I am confused now.

Library extensions are installed like this in Rock in practice:
./lib/x86_64-linux-gnu/ruby/2.0.0/frame_helper_ruby.so
./lib/x86_64-linux-gnu/ruby/2.0.0/typelib_ruby.so
./lib/x86_64-linux-gnu/ruby/2.0.0/base_types_ruby.so
./lib/x86_64-linux-gnu/ruby/2.0.0/jpeg_conversion_ext.so
./lib/x86_64-linux-gnu/ruby/2.0.0/servicediscovery_ruby.so

That means there will be a: "require 'frame_helper_ruby'" or similar
Is that what you would say is good 'Ruby' practice or not?

Right now vizkit would install into:
./lib/x86_64-linux-gnu/ruby/2.0.0/vizkit/vizkittypelib.so

Comment (doudou) [inline modules/RockRuby.cmake]: > Is that what you would say is good 'Ruby' practice or not?

I would try to standardize on the rake-compiler convention, which is also the convention I saw in practice in quite a lot of gems. I.e. having the extension installed under lib/<package_name>/ and required with "<package_name>/extensioname"

Given that extensions are not loaded explicitely - but only through some other ruby files - it should be a change that has no user-visible effect.

**Comment by doudou** [comment]
TL;DR: the rice gem is a hack. Unfortunately, your patch is not multiarch-clean, but I guess that should do for now. Sorry to have wasted your time ...

**Comment by 2maz** [comment]
Right. For now it does the job and no worries, I am happy you have a look at it.

File: modules/RockRuby.cmake (hunk 1 above)
Comment (2maz) [inline modules/RockRuby.cmake]: K, your 'package_name' was not displayed -- now I can follow.
I am agree then with having that as a default and not having a flag, however this would still be a breaking change for libraries with extensions since you have to adapt the require statements?

Comment (2maz) [inline modules/RockRuby.cmake]: Do you have any suggestion on how to proceed here towards (enforcing) a standardized binding layout?

Comment (doudou) [inline modules/RockRuby.cmake]: > I agree with having that as a default and not having a flag, however this would still be a breaking change for libraries with extensions since you have to adapt the require statements?

Indeed ... and unfortunately extensions like frame_helper are required directly (there is no .rb file that does the loading, which we could adapt)... So back to your original idea :(

**Comment by jmachowinski** [comment]
So merge ?

**Comment by 2maz** [comment]
I am all for it @doudou; still objections?

**Comment by doudou** [comment]
nope

**Comment by 2maz** [comment]
Merged then.
